### PR TITLE
[Huya] Add new extractor

### PR DESCRIPTION
**Pull request by hatienl0i261299**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This PR is new extractor to extract video live streaming from `https://www.huya.com/`

Issue: https://github.com/yt-dlp/yt-dlp/issues/2683

### CODE REVIEW

File: yt_dlp/extractor/huya.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        webpage = self._download_webpage(url, video_id=video_id)
+
+        json_stream = self._search_regex(r'"stream": "([a-zA-Z0-9+=/]+)"', webpage, 'stream', default=None)
+
+        if not json_stream:
+            raise ExtractorError('Video is offline')
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:43]:
```suggestion
            raise ExtractorError('Video is offline', expected=True)
```

File: yt_dlp/extractor/huya.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+        stream_data = json.loads(base64.b64decode(json_stream).decode())
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:45]: Use `self._parse_json`

File: yt_dlp/extractor/huya.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        room_info = try_get(stream_data, lambda x: x['data'][0]['gameLiveInfo']) or None
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:46]: why `or None`?

File: yt_dlp/extractor/huya.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+        if not room_info:
+            raise ExtractorError('Can not extract the room info')
+
+        title = room_info.get('introduction', None)
+        if not title:
+            title = self._html_search_regex(r'<title>([^<]+)</title>', webpage, 'title')
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:53]:
```suggestion
        title = room_info.get('introduction') or self._html_search_regex(r'<title>([^<]+)</title>', webpage, 'title')
```

File: yt_dlp/extractor/huya.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        screen_type = room_info.get('screenType')
+        live_source_type = room_info.get('liveSourceType')
+
+        stream_info_list = try_get(stream_data, lambda x: x['data'][0]['gameStreamInfoList'])
+        random.shuffle(stream_info_list)
+        random.shuffle(stream_info_list)
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:59]: why?

File: yt_dlp/extractor/huya.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        stream_info = {}
+        s_url = ''
+        while stream_info_list:
+            stream_info = stream_info_list.pop()
+            s_url = stream_info.get('sFlvUrl')
+            if s_url:
+                break
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:66]: use `next`

File: yt_dlp/extractor/huya.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+        s_stream_name = stream_info.get('sStreamName')
+        s_url_suffix = stream_info.get('sFlvUrlSuffix')
+        _url = f'{s_url}/{s_stream_name}.{s_url_suffix}?'
+
+        re_secret = not screen_type and live_source_type in (0, 8, 13)
+        params = dict(parse_qsl(unescape(stream_info['sFlvAntiCode'])))
+
+        if re_secret:
+            params.setdefault('t', '100')  # 102
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:76]: 102?

File: yt_dlp/extractor/huya.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            u3 = uid & 0xffffff
+            u = u1 | u2 >> 24 | u3 << 8
+            params.update({
+                'u': str(u),
+                'seqid': str(int(ct * 1000) + uid),
+                'ver': '1',
+                'uuid': int(ct % 1e7 * 1e6 % 0xffffffff),
+            })
+            fm = base64.b64decode(params['fm']).decode().split('_', 1)[0]
+            ss = hashlib.md5('|'.join([params.get('seqid'), params.get('ctype'), params.get('t')]))
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:94]: makes sense to move this in a function

File: yt_dlp/extractor/huya.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+        formats = []
+        for si in stream_data['vMultiStreamInfo']:
+            rate = si['iBitRate']
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:98]:
```suggestion
            rate = si.get('iBitRate')
```

File: yt_dlp/extractor/huya.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+                params['ratio'] = rate
+            else:
+                params.pop('ratio', None)
+            if re_secret:
+                params['wsSecret'] = hashlib.md5(
+                    '_'.join([fm, params.get('u'), s_stream_name, ss, params.get('wsTime')]))
+            url = _url + urlencode(params, safe='*')
+            formats.append({
+                'ext': 'flv',
+                'format_id': si['sDisplayName'],
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:109]:
```suggestion
                'format_id': si.get('sDisplayName'),
```

File: yt_dlp/extractor/huya.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+                'url': url,
+                'video_ext': 'flv'
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:111]:
```suggestion
                'ext': 'flv'
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/huya.py (hunk 11 above)
Comment (fstirlitz) [reply yt_dlp/extractor/huya.py:111]: That one is already there.

File: yt_dlp/extractor/huya.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            })
+
+        return {
```
Comment (pukkandan) [inline yt_dlp/extractor/huya.py:114]: do `self._sort_formats(formats)` before returning

File: yt_dlp/extractor/huya.py (hunk 1 above)
Comment (fstirlitz) [reply yt_dlp/extractor/huya.py:43]: Ever thought about making it a separate exception class, say something like `ServiceDeclinedError`?

File: yt_dlp/extractor/huya.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError,
+    try_get
+)
+
+
+class HuyaIE(InfoExtractor):
+    _VALID_URL = r'(?x)^((?:http[s]?|fpt):)\/?\/(?:www\.|m\.|)huya\.com\/(?P<id>.*?)(?:\/|$)'
```
Comment (Lesmiscore) [inline yt_dlp/extractor/huya.py:16]:
```suggestion
    _VALID_URL = r'https?://(?:www\.|m\.)?huya\.com/(?P<id>[^/#?&])'
```
Keep regex clean.

(I'll partially withdraw this comment if you add tests for `fpt://` urls)

File: yt_dlp/extractor/huya.py (hunk 1 above)
Comment (pukkandan) [reply yt_dlp/extractor/huya.py:43]: It makes sense when designing the interface, but I don't see any advantage in adding it now

**Pull request closed without merging**